### ISSUE TITLE
Rails/UniqBeforePluck broken for uniq with block and ActiveRecord::Associations::CollectionProxy

### ISSUE THREAD

**Issue by janraasch** [bbatsov/rubocop#3122]
Take this code

```ruby
Model.where(...).pluck(:id).uniq { |m| some_ruby_logic(m) }
```

## Expected behavior

Let it be. We cannot pull the block to the database.

## Actual behavior

It is marked as an error, plus it is corrected to

```ruby
Model.where(...).uniq.pluck(:id) { |m| some_ruby_logic(m) }
```

## Steps to reproduce the problem

Run `rubocop` on the example given above to see that it errors. Run `rubocop -a` to see the erroneous correction outline above.

## RuboCop version

```
$ rubocop -V
0.40.0 (using Parser 2.3.1.0, running on ruby 2.0.0 x86_64-darwin12.4.0)
```

(The Rails/UniqBeforePluck was introduced with #3111)

**Comment by janraasch** [comment]
Just run into another problem with this Cop:

Perhaps even more problematic is the fact that this does not play nice with `ActiveRecord::Associations::CollectionProxy`.

Take e.g.

```ruby
class Order < ActiveRecord::Base
  has_many :order_items
end

class OrderItem < ActiveRecord::Base
  belongs_to :order
end
```

then

```ruby
o = Order.first
uniq_types = o.order_items.pluck(:type).uniq
```

works as expected, but

```ruby
o = Order.first
uniq_types = o.order_items.uniq.pluck(:type)
```

will throw

```bash
NoMethodError: undefined method `pluck' for #<Array:0x007f9e708b3b28>
```

**Comment by aripollak** [comment]
Perhaps the cop should only work on `distinct`, which is a synonym for `uniq` but doesn't also show up in Array?

**Comment by tobypinder** [comment]
I'm experiencing the same as @janraasch. I'm having problems creating an isolated test case, but I have some pretty complex `ActiveRecord::Relation`s with multiple joins, selects, conditions etc and any attempt in Rails to use `distinct`/`uniq` on this Relation fails as the query is malformed. I apologise for the lack of specific information here and if I get the opportunity I will follow up with minimal test cases, but the long and short of it is that I don't think Rubocop's static analysis will be capable of picking up these or not.

Whether I am an edge case or not I don't know, but if it's affecting others too I would have to recommend disabling the cop by default.

**Comment by tejasbubane** [comment]
The first of the two problems here is fixed by #3125. The second has a separate issue #3148.
So I think we should move the conversation to #3148 and close this one. @bbatsov

**Comment by janraasch** [comment]
I agree. 👍
